rmw_zenoh_cpp/src/rmw_zenoh.cpp

```diff
@@ -1630,145 +1662,695 @@ rmw_return_loaned_message_from_subscription(
+  z_drop(z_move(client_data->zn_closure_reply));
+  z_drop(z_move(client_data->keyexpr));
+  for (z_owned_reply_t & reply : client_data->replies) {
+    z_reply_drop(&reply);
+  }
+  client_data->replies.clear();
+
+  allocator->deallocate(client_data->request_type_support, allocator->state);
+  allocator->deallocate(client_data->response_type_support, allocator->state);
+  allocator->deallocate(client->data, allocator->state);
+
+  allocator->deallocate(const_cast<char *>(client->service_name), allocator->state);
+  allocator->deallocate(client, allocator->state);
+
+  return RMW_RET_OK;
+}
+
+//==============================================================================
+/// Send a ROS service request.
+rmw_ret_t
+rmw_send_request(
+  const rmw_client_t * client,
+  const void * ros_request,
+  int64_t * sequence_id)
+{
+  RCUTILS_LOG_INFO_NAMED(
+    "rmw_zenoh_cpp", "[rmw_send_request]");
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(client->data, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(ros_request, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(sequence_id, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    client,
+    client->implementation_identifier,
+    rmw_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
+
+  rmw_client_data_t * client_data = static_cast<rmw_client_data_t *>(client->data);
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+    client_data,
+    "Unable to retrieve client_data from client.",
+    RMW_RET_INVALID_ARGUMENT);
+
+  rmw_context_impl_s * context_impl = static_cast<rmw_context_impl_s *>(
+    client_data->context->impl);
+
+  // Serialize data
+
+  rcutils_allocator_t * allocator = &(client_data->context->options.allocator);
+
+  size_t max_data_length = (
+    client_data->request_type_support->get_estimated_serialized_size(
+      ros_request, client_data->request_type_support_impl));
+
+  // Init serialized message byte array
+  char * request_bytes = static_cast<char *>(allocator->allocate(
+      max_data_length,
+      allocator->state));
+  if (!request_bytes) {
+    RMW_SET_ERROR_MSG("failed allocate request message bytes");
+    allocator->deallocate(request_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  // Object that manages the raw buffer
+  eprosima::fastcdr::FastBuffer fastbuffer(request_bytes, max_data_length);
+
+  // Object that serializes the data
+  eprosima::fastcdr::Cdr ser(
+    fastbuffer,
+    eprosima::fastcdr::Cdr::DEFAULT_ENDIAN,
+    eprosima::fastcdr::Cdr::DDS_CDR);
+  if (!client_data->request_type_support->serialize_ros_message(
+      ros_request,
+      ser,
+      client_data->request_type_support_impl))
+  {
+    allocator->deallocate(request_bytes, allocator->state);
+    return RMW_RET_ERROR;
+  }
+
+  size_t data_length = ser.getSerializedDataLength();
+
+  // TODO(francocipollone): Do I really need the sequency number here?
+  *sequence_id = 0;
```

Comment:
Hey @clalancette @Yadunund  Do you have more insight about how the sequence id is being used? I imagine this comes from pub/sub implementation for the services which might not be necessary in our case.

Reply:
So I did some research about sequence_id in ROS 2.

First of all, it is defined in https://github.com/ros2/rmw/blob/9b3d9d0e3021b7a6e75d8886e3e061a53c36c789/rmw/include/rmw/rmw.h#L2039-L2045 .

Looking through the implementations, it ends up being used mostly by the RMW layer.  But there, it is always transmitted as part of the service call, so that the service server knows the sequence number hte client chose.  It is then also exposed to the higher layers via `rmw_service_info_t` and `rmw_client_info_t`.  Via that mechanism, it is used in `rcl` when doing service introspection.  I can see this being useful, even in the rmw_zenoh context, to be able to match clients and services together.

Unfortunately, how to generate and store the sequence number is under-specified.  rmw_cyclonedds, for instance, just has a global variable that gets incremented every time a service request is made.  I'm going to suggest that we do something similar here to start with; eventually maybe we want to consider tying this to the node or the session instead of having it global, but we can do it later.

Reply:
If I may, I have 2 suggestions then:
1. make this sequence number optional in the request payload (e.g. generating a local one in rmw_zenoh on server side if not present). Thus a zenoh-pico client or server on a microcontroller is not required to deal with such sequence number.
2. add this sequence number as an attachment to the Zenoh query and reply, rather than piggyback it with the request and reply payloads. This will simplify the serialization and probably avoid a buffer copy.
  You can see an example of attachments usage with pub/sub in [zenoh-c/examples](https://github.com/eclipse-zenoh/zenoh-c/tree/0.10.1-rc/examples). The usage with query/reply is similar (an `attachment` in options struct for query and reply)

Reply:
OK, my latest pushes implement this using attachments, and it seems to work.  There are a couple of other pieces of metadata that we should fill in as well, but we can address those separately.  I'll open an issue about it.